src/handler/active_requests.rs

```diff
@@ -92,12 +120,24 @@ impl ActiveRequests {
 impl Stream for ActiveRequests {
     type Item = Result<(NodeAddress, RequestCall), String>;
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        match self.active_requests_mapping.poll_next_unpin(cx) {
-            Poll::Ready(Some(Ok((node_address, request_call)))) => {
-                // Remove the associated nonce mapping.
-                self.active_requests_nonce_mapping
-                    .remove(request_call.packet().message_nonce());
-                Poll::Ready(Some(Ok((node_address, request_call))))
+        match self.active_requests_nonce_mapping.poll_next_unpin(cx) {
+            Poll::Ready(Some(Ok((nonce, node_address)))) => {
+                match self.active_requests_mapping.entry(node_address.clone()) {
+                    Entry::Vacant(_) => Poll::Ready(None),
+                    Entry::Occupied(mut requests) => {
+                        match requests
+                            .get()
+                            .iter()
+                            .position(|req| req.packet().message_nonce() == &nonce)
+                        {
+                            Some(index) => Poll::Ready(Some(Ok((
+                                node_address,
+                                requests.get_mut().remove(index),
+                            )))),
+                            None => Poll::Ready(None),
+                        }
```

Comment:
> this will leave requests with empty lists

Not sure I'm understanding your suggestion here. Do you want to see an empty requests replaced with `Entry::Vacant` after removing the last request? Based on your suggestion on line 89, it seems like the convention is that it is possible for the `active_requests_mapping` entry for a node to be an empty list instead of strictly `Entry::Vacant`.

In the case where `requests` is an empty list, this logic will simply return `None => Poll::Ready(None)`. There doesn't seem to be a need to handle the empty list case any differently? But I could be missing something.

But if you'd like to see the change where the `active_requests_mapping` entry for a node without any active requests **cannot** be an empty list and **must** always be `Entry::Vacant`, I can make those changes.

Reply:
To be clear, on empty list use `entry.remove()`